src/controller/CHIPDeviceController.cpp

```diff
@@ -1026,6 +1029,105 @@ CHIP_ERROR DeviceCommissioner::UnpairDevice(NodeId remoteDeviceId)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR DeviceCommissioner::SendEchoRequest(NodeId remoteDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Messaging::SendFlags sendFlags;
+    System::PacketBufferHandle payloadBuf;
+    char requestData[kEchoMsgBufSize];
+
+    uint32_t size = mPingParams.GetEchoReqSize();
+    VerifyOrExit(size >= kEchoMsgBufSize, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    snprintf(requestData, kEchoMsgBufSize, "Echo Message %" PRIu32 "\n", mPingParams.GetEchoReqCount() + 1);
+
+    payloadBuf = MessagePacketBuffer::NewWithData(requestData, size);
+    VerifyOrExit(!payloadBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    if (mPingParams.IsUsingMRP())
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNone);
+    }
+    else
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNoAutoRequestAck);
+    }
+
+    mPingParams.SetLastEchoTime(System::Timer::GetCurrentEpoch());
+
+    ChipLogDetail(Controller, "Send echo request message with payload size: %d bytes to Node: 0x" ChipLogFormatX64, size,
+                  ChipLogValueX64(remoteDeviceId));
+
+    mPingParams.SetWaitingForEchoResp(false);
```

Comment:
Shouldn't this set to _true_?

Reply:
why we need to wait for the resp before we send the request?

Reply:
Because in the current state of the tree the response might come back before the send call returns, and then this logic will be broken.  So as a workaround we can set the waiting flag to true first, then set it to false if send fails, which will be more robust to the ordering stuff...

Reply:
Now, I got your point